src/server/scripts/Outland/zone_hellfire_peninsula.cpp

```diff
@@ -390,6 +390,124 @@ class npc_fel_guard_hound : public CreatureScript
+    SPELL_FROSTNOVA = 11831,
+};
+
+class npc_magister_aledis : public CreatureScript
+{
+public:
+    npc_magister_aledis() : CreatureScript("npc_magister_aledis") { }
+
+    struct npc_magister_aledisAI : public ScriptedAI
+    {
+        npc_magister_aledisAI(Creature* creature) : ScriptedAI(creature) { }
+
+        void StartFight(Player* player)
+        {
+            me->Dismount();
+            me->SetFacingToObject(player);
+            me->RemoveNpcFlag(UNIT_NPC_FLAG_GOSSIP);
+            _playerGUID = player->GetGUID();
+            _events.ScheduleEvent(EVENT_TALK, 2s);
+        }
+
+        void Reset() override
+        {
+            me->RestoreFaction();
+            me->RemoveNpcFlag(UNIT_NPC_FLAG_QUESTGIVER);
+            me->SetNpcFlag(UNIT_NPC_FLAG_GOSSIP);
+            me->SetImmuneToPC(true);
+        }
+
+        void DamageTaken(Unit* /*attacker*/, uint32& damage, DamageEffectType /*damageType*/, SpellSchoolMask /*spellInfo = nullptr*/) override
+        {
+            if (damage > me->GetHealth() || me->HealthBelowPctDamaged(20, damage))
+            {
+                damage = 0;
+
+                _events.Reset();
+                me->RestoreFaction();
+                me->RemoveAllAuras();
+                me->GetThreatMgr().ClearAllThreat();
+                me->CombatStop(true);
+                me->SetNpcFlag(UNIT_NPC_FLAG_QUESTGIVER);
+                me->SetImmuneToPC(true);
+                Talk(SAY_DEFEATED);
+
+                _events.ScheduleEvent(EVENT_EVADE, 1min);
+            }
+        }
+
+        void UpdateAI(uint32 diff) override
+        {
+            _events.Update(diff);
+
+            while (uint32 eventId = _events.ExecuteEvent())
+            {
+                switch (eventId)
+                {
+                case EVENT_TALK:
+                    Talk(SAY_CHALLENGE);
+                    _events.ScheduleEvent(EVENT_ATTACK, 2s);
+                    break;
+                case EVENT_ATTACK:
+                    me->SetImmuneToPC(false);
+                    me->SetFaction(FACTION_MONSTER);
+                    if (Player* player = ObjectAccessor::GetPlayer(*me, _playerGUID))
+                    {
+                        AttackStart(player);
+                    }
+                    _events.ScheduleEvent(EVENT_FIREBALL, 1ms);
+                    _events.ScheduleEvent(EVENT_FROSTNOVA, 5s);
+                    break;
+                case EVENT_FIREBALL:
+                    DoCast(SPELL_FIREBALL);
+                    _events.ScheduleEvent(EVENT_FIREBALL, 10s);
+                    break;
+                case EVENT_FROSTNOVA:
+                    DoCastAOE(SPELL_FROSTNOVA);
+                    _events.ScheduleEvent(EVENT_FROSTNOVA, 20s);
+                    break;
+                case EVENT_EVADE:
+                    EnterEvadeMode();
+                    break;
+                }
+            }
+
+            if (!UpdateVictim())
+                return;
```

Comment:
Uhm, maybe move this above the events, and if it's needed add another eventMap for non-combat events (or you could use TaskScheduler.

Reply:
Changed this slightly
```cpp
if (UpdateVictim())
{
    DoMeleeAttackIfReady();
}
```

Reply:
@Nefertumm do you see any issues with the latest changes? Is the rationale clear?

Reply:
If it works, then it's fine for me :)